File: db0/cdb2/calPixelQualityLM.cc
#include "cdbUtil.hh"

#include <iostream>
#include <sstream>
#include "calPixelQualityLM.hh"

using namespace std;

// ----------------------------------------------------------------------
calPixelQualityLM::calPixelQualityLM(cdbAbs *db) : calAbs(db) {
}


// ----------------------------------------------------------------------
string calPixelQualityLM::statusToString(Status status) {
  // -- Update this when adding new Status values
  switch (status) {
    case ChipNotFound: return "ChipNotFound";
    case Good: return "Good";
    case Noisy: return "Noisy";
    case Suspect: return "Suspect";
    case DeclaredBad: return "DeclaredBad";
    case LVDSErrorLink: return "LVDSErrorLink";
    case LVDSErrorOtherLink: return "LVDSErrorOtherLink";
    case LVDSErrorTopBottomEdge: return "LVDSErrorTopBottomEdge";
    case DeadChip: return "DeadChip";
    case NoHits: return "NoHits";
    case Masked: return "Masked";
    default: return "Unknown";
  }
}

// ----------------------------------------------------------------------
string calPixelQualityLM::getStatusDocumentation() {
  // -- List of all Status enum values - update this when adding new Status values
  const Status allStatuses[] = {
    ChipNotFound,
    Good,
    Noisy,
    Suspect,
    DeclaredBad,
    LVDSErrorLink,
    LVDSErrorOtherLink,
    LVDSErrorTopBottomEdge,
    DeadChip,
    NoHits,
    Masked
  };
  
  stringstream ss;
  bool first = true;
  for (Status s : allStatuses) {
    if (!first) ss << ", ";
    ss << static_cast<int>(s) << "=" << statusToString(s);
    first = false;
  }
  ss << ". M-link=nhit/ovfl";
  return ss.str();
}

// ----------------------------------------------------------------------
bool calPixelQualityLM::getNextID(uint32_t &ID) {
  if (fMapConstantsIt == fMapConstants.end()) {
    // -- reset
    ID = 999999;
    fMapConstantsIt = fMapConstants.begin();
    return false;
  } else {
    ID = fMapConstantsIt->first;
    fMapConstantsIt++;
  }
  return true;
}


// ----------------------------------------------------------------------
calPixelQualityLM::calPixelQualityLM(cdbAbs *db, string tag) : calAbs(db, tag) {
  if (0) cout << "calPixelQualityLM created and registered with tag ->" << fTag << "<-"
       << endl;
}


// ----------------------------------------------------------------------
calPixelQualityLM::~calPixelQualityLM() {
  for (auto it: fMapConstants) it.second.mpixel.clear();
  fMapConstants.clear();
  cout << "this is the end of calPixelQualityLM with tag ->" << fTag << "<-" << endl;
}


// ----------------------------------------------------------------------
void calPixelQualityLM::calculate(string hash) {
  cout << "calPixelQualityLM::calculate() with "
       << "fHash ->" << hash << "<-";
  fMapConstants.clear();
  string spl = fTagIOVPayloadMap[hash].fBLOB;

  std::vector<char> buffer(spl.begin(), spl.end());
  std::vector<char>::iterator ibuffer = buffer.begin();

  unsigned int header = blob2UnsignedInt(getData(ibuffer));
  cout << " header: " << hex << header << dec;

  int npix(0), ncol(0);
  while (ibuffer != buffer.end()) {
    constants a;
    a.id = blob2UnsignedInt(getData(ibuffer));
    a.ckdivend = blob2UnsignedInt(getData(ibuffer));
    a.ckdivend2 = blob2UnsignedInt(getData(ibuffer));

    a.linkA = blob2UnsignedInt(getData(ibuffer));
    a.linkB = blob2UnsignedInt(getData(ibuffer));
    a.linkC = blob2UnsignedInt(getData(ibuffer));
    a.linkM = blob2UnsignedInt(getData(ibuffer));

    // -- get number of column entries
    ncol = blob2Int(getData(ibuffer));
    if (ncol > 0) { 
      for (int i = 0; i < ncol; ++i) {
        int icol            = blob2Int(getData(ibuffer));
        unsigned int iqual  = blob2UnsignedInt(getData(ibuffer));
        a.mcol.insert({ icol, static_cast<char>(iqual) });
      }
    }

    // -- get number of pixel entries
    npix = blob2Int(getData(ibuffer));
    if (npix > 0) { 
      for (int i = 0; i < npix; ++i) {
        int icol            = blob2Int(getData(ibuffer));
        int irow            = blob2Int(getData(ibuffer));
        unsigned int iqual = blob2UnsignedInt(getData(ibuffer));
        int idx = icol*250 + irow;
        a.mpixel.insert({ idx, static_cast<char>(iqual) });
      }
    }
    fMapConstants.insert(make_pair(a.id, a));
  }
  cout << " inserted " << fMapConstants.size() << " constants" << endl;

  // -- set iterator over all constants to the start of the map
  fMapConstantsIt = fMapConstants.begin();
}


// ----------------------------------------------------------------------
calPixelQualityLM::Status calPixelQualityLM::getColStatus(unsigned int chipid, int icol) {
  if (fMapConstants.find(chipid) == fMapConstants.end()) {
    return Status::ChipNotFound; // -- chip not found
  }
  return static_cast<Status>(fMapConstants[chipid].mcol[icol]);
}

// ----------------------------------------------------------------------
int calPixelQualityLM::getNpixWithStatus(unsigned int chipid, Status status) {
  if (fVerbose > 0) cout << "calPixelQualityLM::getNpixWithStatus> chipid = " << chipid << " status = " << status << endl;
  if (fMapConstants.find(chipid) == fMapConstants.end()) {
    return -1; // -- chip not found
  }
  // -- n counts the defective pixels. If status == 0 (Good) was requested, then it is inverted at the end.
  int n(0);
  // -- check pixel status
  for (auto it: fMapConstants[chipid].mpixel) {
    if (fVerbose > 4) cout << "calPixelQualityLM::getNpixWithStatus> pixel = " << it.first << " status = " << static_cast<int>(it.second) << endl;
    if (0 == status) {
      // -- count all pixels with bad status
      if (it.second != 0) n++;
    } else {
      if (it.second == status) n++;
    }
  }
  // -- check column status
  for (auto it: fMapConstants[chipid].mcol) {
    if (fVerbose > 4) cout << "calPixelQualityLM::getNpixWithStatus> column = " << it.first << " status = " << static_cast<int>(it.second)  << endl;
    if (0 == status){
      // -- count all columns with bad status
      if (it.second != 0) n += 250;
    } else {
      if (it.second == status) n += 250; // -- all pixels in this column are defective
    }
  }

  // -- check link status
  if (0 == status) {
    if (fMapConstants[chipid].linkA != 0) {
      n += 250*89;
    }
    if (fMapConstants[chipid].linkB != 0) {
      n += 250*84;
    }
    if (fMapConstants[chipid].linkC != 0) {
      n += 250*83;
    }
  } else {
    if (fMapConstants[chipid].linkA == status) {
      n += 250*89;
    }
    if (fMapConstants[chipid].linkB == status) {
      n += 250*84;
    }
    if (fMapConstants[chipid].linkC == status) {
      n += 250*83;
    }
  }


  if (0 == status) {
    n = 256*250 - n;
    if (fVerbose > 0) cout << "calPixelQualityLM::getNpixWithStatus> chipID = " << chipid << " has " << n << " pixels with good status" << endl;
    //cout << "chipID = " << chipid << " has " << n << " pixels with good status" << endl;
  } else {
    //cout << "chipID = " << chipid << " has " << n << " pixels with status " << status << endl;
  }

  return n;
}

// ----------------------------------------------------------------------
calPixelQualityLM::Status calPixelQualityLM::getStatus(unsigned int chipid, int icol, int irow) {
  // -- first check dead links  
  if (fMapConstants[chipid].linkA > 0 && icol >= 0 && icol <= 87) {
    return static_cast<Status>(fMapConstants[chipid].linkA);
  } 
  if (fMapConstants[chipid].linkB > 0 && icol >= 88 && icol <= 171) {
    return static_cast<Status>(fMapConstants[chipid].linkB);
  }
  if (fMapConstants[chipid].linkC > 0 && icol >= 172 && icol <= 255) {
    return static_cast<Status>(fMapConstants[chipid].linkC);
  }
  // -- now check  column status
  if (fMapConstants[chipid].mcol.find(icol) != fMapConstants[chipid].mcol.end()) {
    return static_cast<Status>(fMapConstants[chipid].mcol[icol]);
  } 

  // -- finally check pixel status
  if (fMapConstants[chipid].mpixel.find(icol*250+irow) == fMapConstants[chipid].mpixel.end()) {
    return Status::Good;
  } else {
    return static_cast<Status>(fMapConstants[chipid].mpixel[icol*250+irow]);
  }
}

// ----------------------------------------------------------------------
int calPixelQualityLM::getCkdivend(unsigned int chipid) {
  if (fMapConstants.find(chipid) == fMapConstants.end()) {
    return -1; // -- chip not found
  }
  return fMapConstants[chipid].ckdivend;
}

// ----------------------------------------------------------------------
int calPixelQualityLM::getCkdivend2(unsigned int chipid) {
  if (fMapConstants.find(chipid) == fMapConstants.end()) {
    return -1; // -- chip not found
  }
  return fMapConstants[chipid].ckdivend2;
}


// ----------------------------------------------------------------------
bool calPixelQualityLM::isLinkBad(unsigned int chipid, int ilink) {
  if (fMapConstants.find(chipid) == fMapConstants.end()) {
    return true; // -- chip not found
  }
  bool result(false);
  switch (ilink) {
    case 0:
      if (fMapConstants[chipid].linkA > 0) {
        result = true;
        break;
      }
      break;
    case 1:
      if (fMapConstants[chipid].linkB > 0) {
        result = true;
        break;
      }
      break;
    case 2:
      if (fMapConstants[chipid].linkC > 0) {
        result = true;
        break;
      }
      break;
    default:
      result = true; // -- should not happen
      break;
  }
  return result;
}

// ----------------------------------------------------------------------
bool calPixelQualityLM::isLinkDead(unsigned int chipid, int ilink) {
  if (fMapConstants.find(chipid) == fMapConstants.end()) {
    return true; // -- chip not found
  }
  switch (ilink) {
    case 0:
      if (fMapConstants[chipid].linkA == 8 || fMapConstants[chipid].linkA == 9) {
        return true;
      }
      return false;
    case 1:
      if (fMapConstants[chipid].linkB == 8 || fMapConstants[chipid].linkB == 9) {
        return true;
      }
      return false;
    case 2:
      if (fMapConstants[chipid].linkC == 8 || fMapConstants[chipid].linkC == 9) {
        return true;
      }
      return false;
    default:
      return true; // -- should not happen
  }
}

// ----------------------------------------------------------------------
bool calPixelQualityLM::isChipDead(unsigned int chipid, int row, int col) {
  if (fMapConstants.find(chipid) == fMapConstants.end()) {
    return true; // -- chip not found
  }
  int cntDeadLinks(0); 
  for (int ilink = 0; ilink < 3; ++ilink) { 
    if (isLinkDead(chipid, ilink)) {
      cntDeadLinks++;
    }
  }
  return cntDeadLinks == 3;
}


// ----------------------------------------------------------------------
void calPixelQualityLM::printPixelQuality(unsigned int chipid, int minimumStatus) {
  // FIXME
}


// ----------------------------------------------------------------------
void calPixelQualityLM::printBLOB(std::string sblob, int verbosity) {
  cout << printBLOBString(sblob, verbosity) << endl;
}

// ----------------------------------------------------------------------
string calPixelQualityLM::printBLOBString(std::string sblob, int verbosity) {
  stringstream ss;

  std::vector<char> buffer(sblob.begin(), sblob.end());
  std::vector<char>::iterator ibuffer = buffer.begin();

  unsigned int header = blob2UnsignedInt(getData(ibuffer));
  ss << "calPixelQuality::printBLOB(string)" << endl;
  ss << "   header: " << hex << header << dec << endl;

  while (ibuffer != buffer.end()) {
    // -- chipID
    unsigned int chipID = blob2UnsignedInt(getData(ibuffer));
    // -- ckdivend and ckdivend2
    unsigned int ckdivend = blob2UnsignedInt(getData(ibuffer));
    unsigned int ckdivend2 = blob2UnsignedInt(getData(ibuffer));
    // -- get link words
    unsigned int linkA = blob2UnsignedInt(getData(ibuffer));
    unsigned int linkB = blob2UnsignedInt(getData(ibuffer));
    unsigned int linkC = blob2UnsignedInt(getData(ibuffer));
    unsigned int linkM = blob2UnsignedInt(getData(ibuffer));
    ss << "   chipID: " << chipID << " ckdivend/ckdivend2: " << ckdivend << "/" << ckdivend2;
    ss << " link status A/B/C/M: " << linkA  << "/" << linkB << "/" << linkC  << "/" << linkM << endl;
    // -- get number of column entries
    int ncol = blob2Int(getData(ibuffer));
    if (ncol > 0) { 
      ss << "            column status " << ncol << " (col/qual): ";
      for (int i = 0; i < ncol; ++i) {
        int icol           = blob2Int(getData(ibuffer));
        unsigned int iqual = blob2UnsignedInt(getData(ibuffer));
        ss << icol << "/" << iqual << (i < ncol-1? ", ":"");
      }
      ss << endl;
    }
    // -- get number of pixel entries
    int npix = blob2Int(getData(ibuffer));
    if (npix > 0) { 
      ss << "            defective pixels " << npix << " (col/row/qual): " ;
      for (int i = 0; i < npix; ++i) {
        int icol           = blob2Int(getData(ibuffer));
        int irow           = blob2Int(getData(ibuffer));
        unsigned int iqual = blob2UnsignedInt(getData(ibuffer));
        ss << icol << "/" << irow << "/" << iqual << (i < npix-1? ", ":"");
      }
      ss << endl;
    }
  }
  return ss.str();
}


// ----------------------------------------------------------------------
string calPixelQualityLM::makeBLOB() {
  stringstream s;
  unsigned int header(0xdeadface);
  s << dumpArray(uint2Blob(header));

  // -- format of m
  // chipID => [npix, n*(col, row, iqual)]
  for (auto it: fMapConstants) {
    s << dumpArray(uint2Blob(it.first));
    constants a = it.second;

    s << dumpArray(uint2Blob(a.ckdivend)); // -- ckdivend
    s << dumpArray(uint2Blob(a.ckdivend2)); // -- ckdivend2

    s << dumpArray(uint2Blob(a.linkA)); // -- linkA
    s << dumpArray(uint2Blob(a.linkB)); // -- linkB
    s << dumpArray(uint2Blob(a.linkC)); // -- linkC
    s << dumpArray(uint2Blob(a.linkM)); // -- linkM

    // -- get number of column entries
    int ncol = a.mcol.size();
    s << dumpArray(int2Blob(ncol));
    for (auto it: a.mcol) {
      int icol = it.first;
      unsigned int iqual = static_cast<unsigned int>(it.second);
      s << dumpArray(int2Blob(icol));
      s << dumpArray(uint2Blob(iqual));
    }

    // -- get number of pixel entries
    int npix = a.mpixel.size();
    s << dumpArray(int2Blob(npix));
    for (auto it: a.mpixel) {
      int idx = it.first;
      int icol = idx/250;
      int irow = idx%250;
      unsigned int iqual = static_cast<unsigned int>(it.second);
      s << dumpArray(int2Blob(icol));
      s << dumpArray(int2Blob(irow));
      s << dumpArray(uint2Blob(iqual));
    }

  }
  return s.str();

}

// ----------------------------------------------------------------------
string calPixelQualityLM::makeBLOB(const map<unsigned int, vector<double>>& m) {
  stringstream s;
  unsigned int header(0xdeadface);
  s << dumpArray(uint2Blob(header));

  // -- format of m
  // chipID => ckdivend,ckdivend2,linkA,linkB,linkC,linkM,ncol[,icol,iqual],npix[,icol,irow,iqual]
  for (auto it: m) {
    s << dumpArray(uint2Blob(it.first));
    // -- ckdivend and ckdivend2
    s << dumpArray(uint2Blob(31));
    s << dumpArray(uint2Blob(31));
    // -- link words (dummy for now)
    s << dumpArray(uint2Blob(0));
    s << dumpArray(uint2Blob(0));
    s << dumpArray(uint2Blob(0));
    s << dumpArray(uint2Blob(0));
    int npix = it.second.size()/3;
    s << dumpArray(int2Blob(npix));
    for (int ipix = 0; ipix < npix; ++ipix) {
      int idx   = ipix*3;
      int icol  = static_cast<int>(it.second[idx]);
      idx       = ipix*3 + 1;
      int irow  = static_cast<int>(it.second[idx]);
      idx       = ipix*3 + 2;
      int iqual = static_cast<int>(it.second[idx]);

      s << dumpArray(int2Blob(icol));
      s << dumpArray(int2Blob(irow));
      s << dumpArray(int2Blob(iqual));
    }
  }
  return s.str();
}


// ----------------------------------------------------------------------
void calPixelQualityLM::readCsv(string filename) {
  // cout << "calPixelQualityLM::readCsv> reset fMapConstants" << endl;
  fMapConstants.clear();

  string spl("");
  ifstream INS(filename);
  if (!INS.is_open()) {
    cout << ("calPixelQualityLM::readCsv> Error, file "
             + filename + " not found")
         << endl;
  }

  vector<string> vline;
  while (getline(INS, spl)) {
    if (string::npos == spl.find("#")) {
      vline.push_back(spl);
    }
  }
  INS.close();

  for (unsigned int it = 0; it < vline.size(); ++it) {
    constants a;
    vector<string> tokens = split(vline[it], ',');
    // -- chipID
    a.id = stoi(tokens[0]);
    a.ckdivend = stoi(tokens[1]);
    a.ckdivend2 = stoi(tokens[2]);
    a.linkA = stoi(tokens[3]);
    a.linkB = stoi(tokens[4]);
    a.linkC = stoi(tokens[5]);
    a.linkM = stoi(tokens[6]);
    // -- initialize column map
    int ncol = stoi(tokens[7]);
    a.mcol.clear();
    for (unsigned ipix = 0; ipix < ncol; ++ipix) {
      int icol           = stoi(tokens[8 + 2*ipix]);
      unsigned int iqual = stoi(tokens[8 + 2*ipix + 1]);
      a.mcol[icol] = static_cast<char>(iqual);
    }
    // -- initialize pixel map
    int npix = stoi(tokens[8 + 2*ncol]);
    a.mpixel.clear();
    for (unsigned ipix = 0; ipix < npix; ++ipix) {
      int icol           = stoi(tokens[9 + 2*ncol + ipix*3]);
      int irow           = stoi(tokens[9 + 2*ncol + ipix*3 + 1]);
      int iqual          = stoi(tokens[9 + 2*ncol + ipix*3 + 2]);
      a.mpixel[icol*250 + irow] = static_cast<char>(iqual);
    }
    fMapConstants.insert(make_pair(a.id, a));
  }

  // -- set iterator over all constants to the start of the map
  fMapConstantsIt = fMapConstants.begin();
}

// ----------------------------------------------------------------------
void calPixelQualityLM::writeCsv(string filename) {
  ofstream OUTS(filename);
  OUTS << "#chipID,ckdivend,ckdivend2,linkA,linkB,linkC,linkM,ncol[,icol,iqual],npix[,icol,irow,iqual] NB: 0 = good, 1 = noisy, 2 = suspect, 3 = declared bad, 9 = turned off" << endl;

 
  for (auto it: fMapConstants) {
    OUTS << it.first << "," << it.second.ckdivend << "," << it.second.ckdivend2 << "," 
         << it.second.linkA << "," << it.second.linkB << "," << it.second.linkC << "," 
         << it.second.linkM << ",";
    OUTS << it.second.mcol.size();
    for (auto itC: it.second.mcol) {
      OUTS << "," << itC.first;
      OUTS << "," << static_cast<int>(itC.second);
    }
    OUTS << "," << it.second.mpixel.size();
    for (auto itP: it.second.mpixel) {
      OUTS << "," << itP.first/250 << "," << itP.first%250 << "," << static_cast<int>(itP.second);
    }
    OUTS << endl;
  }
  OUTS.close();
}

// ----------------------------------------------------------------------
calPixelQualityLM::Status calPixelQualityLM::getLinkStatus(unsigned int chipid, int ilink) {
  if (fMapConstants.find(chipid) == fMapConstants.end()) {
    return Status::ChipNotFound; // -- chip not found
  }
  
  switch (ilink) {
    case 0: return static_cast<Status>(fMapConstants[chipid].linkA);
    case 1: return static_cast<Status>(fMapConstants[chipid].linkB);
    case 2: return static_cast<Status>(fMapConstants[chipid].linkC);
    case 3: return static_cast<Status>(fMapConstants[chipid].linkM);
    default: return Status::ChipNotFound; // -- invalid link number
  }
}

File: mu3e-cdb-python/mu3e_cdb/calibrations/CalPixelQualityLM.py
"""Pixel quality calibration class."""

from enum import IntEnum
from typing import Optional
from .base import Calibration
from ..database.base import Database


class PixelQualityStatus(IntEnum):
    """Pixel quality status enumeration."""
    CHIP_NOT_FOUND = -1
    GOOD = 0
    NOISY = 1
    SUSPECT = 2
    DECLARED_BAD = 3
    LVDS_ERROR_LINK = 4
    LVDS_ERROR_OTHER_LINK = 5
    LVDS_ERROR_TOP_BOTTOM_EDGE = 6
    DEAD_CHIP = 7
    NO_HITS = 8
    MASKED = 9
    
    @classmethod
    def to_string(cls, status: 'PixelQualityStatus') -> str:
        """Convert status to string."""
        status_map = {
            cls.CHIP_NOT_FOUND: "ChipNotFound",
            cls.GOOD: "Good",
            cls.NOISY: "Noisy",
            cls.SUSPECT: "Suspect",
            cls.DECLARED_BAD: "DeclaredBad",
            cls.LVDS_ERROR_LINK: "LVDSErrorLink",
            cls.LVDS_ERROR_OTHER_LINK: "LVDSErrorOtherLink",
            cls.LVDS_ERROR_TOP_BOTTOM_EDGE: "LVDSErrorTopBottomEdge",
            cls.DEAD_CHIP: "DeadChip",
            cls.NO_HITS: "NoHits",
            cls.MASKED: "Masked",
        }
        return status_map.get(status, "Unknown")
    
    @classmethod
    def get_documentation(cls) -> str:
        """Get documentation string for all status values."""
        all_statuses = [
            cls.CHIP_NOT_FOUND,
            cls.GOOD,
            cls.NOISY,
            cls.SUSPECT,
            cls.DECLARED_BAD,
            cls.LVDS_ERROR_LINK,
            cls.LVDS_ERROR_OTHER_LINK,
            cls.LVDS_ERROR_TOP_BOTTOM_EDGE,
            cls.DEAD_CHIP,
            cls.NO_HITS,
            cls.MASKED,
        ]
        
        doc_parts = [f"{int(s)}={cls.to_string(s)}" for s in all_statuses]
        return ", ".join(doc_parts) + ". M-link=nhit/ovfl"


class CalPixelQualityLM(Calibration):
    """Pixel quality calibration using 3+1 link words (matches C++ calPixelQualityLM)."""
    
    def __init__(self, db: Optional[Database] = None, tag: Optional[str] = None):
        super().__init__(db, tag)
        self._constants = {}  # chip_id -> constants dict
    
    def get_name(self) -> str:
        """Return the calibration name."""
        return "pixelqualitylm_"
    
    def calculate(self, hash: str) -> None:
        """Load and calculate calibration from hash."""
        if not self.db:
            raise ValueError("Database not set")
        
        payload = self.db.get_payload(hash)
        if not payload:
            if self.verbose > 0:
                print(f"Warning: Payload not found for hash: {hash}")
            self.hash = hash
            self._constants.clear()
            return
        
        # Check if BLOB is empty
        if not payload.blob or len(payload.blob.strip()) == 0:
            if self.verbose > 0:
                print(f"Warning: Payload {hash} has empty BLOB")
            self.hash = hash
            self._constants.clear()
            return
        
        self.hash = hash
        self._constants.clear()
        
        # Use schema from payload to decode BLOB dynamically
        schema = payload.schema
        if not schema:
            # Fallback to hardcoded schema if payload doesn't have one
            schema = self.get_schema()
            if self.verbose > 0:
                print(f"Warning: No schema in payload, using default: {schema}")
        
        # Decode base64 BLOB
        from ..utils.blob import decode_blob_string, read_blob_header
        from ..utils.schema import SchemaDecoder
        
        try:
            blob_binary = decode_blob_string(payload.blob)
        except Exception as e:
            raise ValueError(f"Failed to decode BLOB (base64): {e}")
        
        if len(blob_binary) < 8:
            if self.verbose > 0:
                print(f"Warning: BLOB too short: {len(blob_binary)} bytes (need at least 8 for header)")
            return
        
        # Read header
        blob_iter = iter(blob_binary)
        try:
            header = read_blob_header(blob_iter)
            if self.verbose > 0:
                print(f"CalPixelQualityLM::calculate() with hash={hash}, header=0x{header:x}")
                print(f"  Using schema: {schema}")
                print(f"  BLOB size: {len(blob_binary)} bytes")
        except (StopIteration, ValueError) as e:
            raise ValueError(f"Failed to read BLOB header: {e}")
        
        # Create schema decoder
        decoder = SchemaDecoder(schema)
        
        # Parse constants using schema
        record_count = 0
        # Parse constants using schema
        record_count = 0
        while True:
            try:
                # Decode one record using the schema
                decoded = decoder.decode_blob(blob_binary, blob_iter)
                
                # Check if we got valid data (at least id field)
                if "id" not in decoded:
                    if self.verbose > 1:
                        print(f"  No 'id' field in decoded record {record_count + 1}")
                    break
                
                chip_id = decoded["id"]
                record_count += 1
                
                if self.verbose > 0 and record_count <= 15:
                    print(f"  Record {record_count}: chip_id={chip_id}, linkA={decoded.get('linkA')}, linkB={decoded.get('linkB')}, linkC={decoded.get('linkC')}, linkM={decoded.get('linkM')}, ncol={decoded.get('ncol', 'MISSING')}, npix={decoded.get('npix', 'MISSING')}")
                
                # Extract fields
                mcol = {}
                if "ncol" in decoded and decoded["ncol"] > 0:
                    ncol_data = decoded.get("ncol_data", [])
                    if self.verbose > 0 and record_count <= 5:
                        print(f"    ncol_data has {len(ncol_data)} items")
                    for item in ncol_data:
                        if "col" in item and "qual" in item:
                            mcol[item["col"]] = item["qual"]
                
                mpixel = {}
                if "npix" in decoded and decoded["npix"] > 0:
                    npix_data = decoded.get("npix_data", [])
                    if self.verbose > 0 and record_count <= 5:
                        print(f"    npix_data has {len(npix_data)} items (expected {decoded['npix']})")
                    for item in npix_data:
                        if "col" in item and "row" in item and "qual" in item:
                            idx = item["col"] * 250 + item["row"]
                            mpixel[idx] = item["qual"]
                
                try:
                    self._constants[chip_id] = {
                        "ckdivend": decoded.get("ckdivend", 0),
                        "ckdivend2": decoded.get("ckdivend2", 0),
                        "linkA": decoded.get("linkA", 0),
                        "linkB": decoded.get("linkB", 0),
                        "linkC": decoded.get("linkC", 0),
                        "linkM": decoded.get("linkM", 0),
                        "mcol": mcol,
                        "mpixel": mpixel,
                    }
                    if self.verbose > 0 and record_count <= 10:
                        print(f"    Successfully inserted chip {chip_id}")
                except Exception as e:
                    if self.verbose > 0:
                        print(f"    ERROR inserting chip {chip_id}: {e}")
                        import traceback
                        traceback.print_exc()
                    raise
            except StopIteration:
                # End of data - no more bytes available
                if self.verbose > 0:
                    print(f"  Reached end of BLOB data after {record_count} records")
                break
            except (KeyError, ValueError) as e:
                # Parsing error
                if self.verbose > 0:
                    print(f"  Error parsing record {record_count + 1}: {e}")
                    import traceback
                    traceback.print_exc()
                break
            except Exception as e:
                # Unexpected error
                if self.verbose > 0:
                    print(f"  Unexpected error parsing record {record_count + 1}: {e}")
                    import traceback
                    traceback.print_exc()
                break
        
        if self.verbose > 0:
            print(f"CalPixelQualityLM::calculate() inserted {len(self._constants)} constants")
            if len(self._constants) > 0:
                # Print first few chip IDs
                chip_ids = sorted(self._constants.keys())[:10]
                print(f"  Sample chip IDs: {chip_ids}")
                if 36 in self._constants:
                    chip36 = self._constants[36]
                    print(f"  Chip 36 found: linkA={chip36.get('linkA')}, linkB={chip36.get('linkB')}, linkC={chip36.get('linkC')}, linkM={chip36.get('linkM')}")
                else:
                    print(f"  Chip 36 NOT found in decoded constants!")
    
    def get_schema(self) -> str:
        """Return the schema string."""
        return "ui_id,ui_ckdivend,ui_ckdivend2,ui_linkA,ui_linkB,ui_linkC,ui_linkM,i_ncol[,i_col,ui_qual],i_npix[,i_col,i_row,ui_qual]"
    
    def get_status(self, chip_id: int, col: int, row: int) -> PixelQualityStatus:
        """Get pixel status for a given chip, column, and row."""
        if chip_id not in self._constants:
            return PixelQualityStatus.CHIP_NOT_FOUND
        
        const = self._constants[chip_id]
        
        # First check dead links (matches C++ implementation)
        if const["linkA"] > 0 and 0 <= col <= 87:
            return PixelQualityStatus(const["linkA"])
        if const["linkB"] > 0 and 88 <= col <= 171:
            return PixelQualityStatus(const["linkB"])
        if const["linkC"] > 0 and 172 <= col <= 255:
            return PixelQualityStatus(const["linkC"])
        
        # Now check column status
        if col in const["mcol"]:
            return PixelQualityStatus(const["mcol"][col])
        
        # Finally check pixel status
        idx = col * 250 + row
        if idx in const["mpixel"]:
            return PixelQualityStatus(const["mpixel"][idx])
        
        # Default to good
        return PixelQualityStatus.GOOD
    
    def get_col_status(self, chip_id: int, col: int) -> PixelQualityStatus:
        """Get column status for a given chip and column."""
        if chip_id not in self._constants:
            return PixelQualityStatus.CHIP_NOT_FOUND
        
        const = self._constants[chip_id]
        if col in const["mcol"]:
            return PixelQualityStatus(const["mcol"][col])
        
        return PixelQualityStatus.GOOD
    
    def get_link_status(self, chip_id: int, link: int) -> PixelQualityStatus:
        """Get link status for a given chip and link (0-3)."""
        if chip_id not in self._constants:
            return PixelQualityStatus.CHIP_NOT_FOUND
        
        const = self._constants[chip_id]
        link_map = {0: "linkA", 1: "linkB", 2: "linkC", 3: "linkM"}
        if link in link_map:
            link_value = const[link_map[link]]
            if link_value == 0:
                return PixelQualityStatus.GOOD
            return PixelQualityStatus(link_value)
        
        return PixelQualityStatus.CHIP_NOT_FOUND
    
    def is_chip_dead(self, chip_id: int, row: int = -1, col: int = -1) -> bool:
        """Check if chip is dead (all three links are dead: DeadChip, NoHits, or Masked)."""
        if chip_id not in self._constants:
            return True
        
        # Check if all three links are dead (status 7, 8, or 9)
        cnt_dead_links = 0
        for ilink in range(3):
            if self.is_link_dead(chip_id, ilink):
                cnt_dead_links += 1
        return cnt_dead_links == 3
    
    def is_link_bad(self, chip_id: int, link: int) -> bool:
        """Check if link is bad."""
        status = self.get_link_status(chip_id, link)
        return status != PixelQualityStatus.GOOD
    
    def is_link_dead(self, chip_id: int, link: int) -> bool:
        """Check if link is dead (DeadChip=7, NoHits=8, or Masked=9)."""
        status = self.get_link_status(chip_id, link)
        return status in (PixelQualityStatus.DEAD_CHIP, PixelQualityStatus.NO_HITS, PixelQualityStatus.MASKED)
    
    def get_lvds_overflow_rate(self, chip_id: int) -> float:
        """Get LVDS overflow rate for a chip."""
        if chip_id not in self._constants:
            return -1.0
        
        const = self._constants[chip_id]
        link_m = const.get("linkM", 0)
        
        if link_m == 0:
            return 0.0  # No overflow
        elif link_m == 1:
            return 9.99e99  # Overflow rate = 0 (sentinel value)
        else:
            return 1.0 / float(link_m)
    
    def get_ckdivend(self, chip_id: int) -> int:
        """Get ckdivend for a chip."""
        if chip_id not in self._constants:
            return -1
        return self._constants[chip_id].get("ckdivend", 0)
    
    def get_ckdivend2(self, chip_id: int) -> int:
        """Get ckdivend2 for a chip."""
        if chip_id not in self._constants:
            return -1
        return self._constants[chip_id].get("ckdivend2", 0)
    
    def get_npix_with_status(self, chip_id: int, status: PixelQualityStatus) -> int:
        """Get number of pixels with given status."""
        if chip_id not in self._constants:
            return 0
        
        # Implementation would count pixels with given status
        return 0

